cvmfs/network/download.cc

```diff
@@ -1880,23 +2067,105 @@ Failures DownloadManager::Fetch(JobInfo *info) {
-    } while (true);
+        switch (ele->action) {
+          case kActionStop:
+            is_running = false;
+          break;
+          case kActionEndOfData:
+          {
+            const int curl_error = *ele->data;
+            const bool success = VerifyDownloadSuccess(curl_error, info);
+
+            DataTubeElement *out_ele = info->parallel_dwnld_coord()->
+                                                     GetUnusedDataTubeElement();
+
+            if (success) {
+              out_ele->action = kActionDownloadDone;
+            } else {
+              out_ele->action = kActionCheckRepeat;
+            }
+            info->GetCmdTubePtr()->EnqueueBack(out_ele);
+          }
+          break;
+          case kActionShouldRetry:
+          {
+            const bool retry = ShouldRetry(info);
+
+            DataTubeElement *out_ele = info->parallel_dwnld_coord()->
+                                                     GetUnusedDataTubeElement();
+            if (retry) {
+              out_ele->action = kActionRetry;
+            } else {
+              out_ele->action = kActionDownloadDone;
+            }
+            info->GetCmdTubePtr()->EnqueueBack(out_ele);
+          }
+          break;
+          case kActionData:
+          {
+            // quick escape
+            if (info->stop_data_download()) {
+              break;
+            }
+
+            char *ptr = ele->data;
+            const size_t num_bytes = ele->size;
+            if (info->expected_hash()) {
+              shash::Update(reinterpret_cast<unsigned char *>(ptr),
+                            num_bytes, info->hash_context());
+            }
+
+            if (info->compressed()) {
+              const zlib::StreamStates retval =
+                zlib::DecompressZStream2Sink(ptr,
+                                           static_cast<int64_t>(num_bytes),
+                                           info->GetZstreamPtr(), info->sink());
+              if (retval == zlib::kStreamDataError) {
+                LogCvmfs(kLogDownload, kLogSyslogErr,
+                                     "(id %" PRId64 ") failed to decompress %s",
+                                     info->id(), info->url()->c_str());
+                info->SetErrorCode(kFailBadData);
+                info->SetStopDataDownload(true);
+              } else if (retval == zlib::kStreamIOError) {
+                LogCvmfs(kLogDownload, kLogSyslogErr,
+                            "(id %" PRId64 ") decompressing %s, local IO error",
+                            info->id(), info->url()->c_str());
+                info->SetErrorCode(kFailLocalIO);
+                info->SetStopDataDownload(true);
+              }
+            } else {
+              const int64_t written = info->sink()->Write(ptr, num_bytes);
+              if (written < 0 || static_cast<uint64_t>(written) != num_bytes) {
+                PANIC(kLogStderr | kLogDebug, "(id %" PRId64 ") "
```

Comment:
In the `CallbackCurlData()` this is a plain `LogCvmfs()` statement, but here its a `PANIC()`. Is there a reason why? If possible, it might make sense to refactor these section into a single helper if they are identical. Also, if we cant write to the sync, we should stop the download, right?